lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -478,13 +512,31 @@ void FIRRTLTypesLowering::visitStmt(ConnectOp op) {
   assert(destValues.size() == srcValues.size() &&
          "connected bundles don't match");
 
+  // Determine if the LHS expression is the duplex value.
+  auto isDestDuplex = isDuplexValue(destValues.front());
+
   for (auto tuple : llvm::zip_first(destValues, srcValues)) {
     Value newDest = std::get<0>(tuple);
     Value newSrc = std::get<1>(tuple);
-    if (newDest.getType().isa<FlipType>())
-      builder->create<ConnectOp>(newDest, newSrc);
-    else
-      builder->create<ConnectOp>(newSrc, newDest);
+
+    // The rules for flow state that a value from a duplex expression can be
+    // used as both a source and sink, regardless of the flip orientation of the
+    // type. To make this work, we find the non-duplex value and make sure that
+    // it is in the correct position. Two duplex values cannot be connected,
+    // since it is unclear which side is left or right.
+    if (isDestDuplex) {
```

Comment:
Nice! This piece of code always felt fragile... I think this is much more solid.